### PR TITLE
[4.x only] Lazy Loading events

### PR DESCRIPTION
**Pull request by vicb**
supersedes #14232

/cc @DzmitryShylovich

### CODE REVIEW

File: modules/@angular/router/src/events.ts [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+  constructor(public route: Route) {}
+
+  toString(): string { return `RouteConfigLoadEnd(path: ${this.route.path})`; }
+}
+
+/**
+ * @whatItDoes Represents a router event.
+ *
+ * Please see {@link NavigationStart}, {@link NavigationEnd}, {@link NavigationCancel}, {@link
+  * NavigationError}, {@link RoutesRecognized}, {@link RouteConfigLoadEnd} for more information.
```
Comment (DzmitryShylovich) [inline modules/@angular/router/src/events.ts]: `+ RouteConfigLoadStart`

File: modules/@angular/router/src/events.ts [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+  }
+}
+
+/**
+ * @whatItDoes Represents an event triggered when routes are recognized.
+ *
+ * @stable
+ */
+export class RoutesRecognized {
+  // TODO: vsavkin: make internal
```
Comment (DzmitryShylovich) [inline modules/@angular/router/src/events.ts]: maybe you should fix these todos or remove them?

File: modules/@angular/router/src/apply_redirects.ts [hunk 3]
```diff
@@ -219,80 +226,80 @@ class ApplyRedirects {
...
+    });
   }
 
   private matchSegmentAgainstRoute(
       injector: Injector, rawSegmentGroup: UrlSegmentGroup, route: Route,
       segments: UrlSegment[]): Observable<UrlSegmentGroup> {
     if (route.path === '**') {
       if (route.loadChildren) {
-        return map.call(this.configLoader.load(injector, route.loadChildren), (r: any) => {
+        return map.call(this.configLoader.load(injector, route), (r: any) => {
```
Comment (DzmitryShylovich) [inline modules/@angular/router/src/apply_redirects.ts]: `(r: any)` => `(c: LoadedRouterConfig)`

File: modules/@angular/router/src/apply_redirects.ts [hunk 4]
```diff
@@ -219,80 +226,80 @@ class ApplyRedirects {
...
-      });
+      return of (new UrlSegmentGroup(segments, {}));
     }
+
+    const {matched, consumedSegments, lastChild} = match(rawSegmentGroup, route, segments);
+    if (!matched) return noMatch(rawSegmentGroup);
+
+    const rawSlicedSegments = segments.slice(lastChild);
+    const childConfig$ = this.getChildConfig(injector, route);
+    return mergeMap.call(childConfig$, (routerConfig: any) => {
```
Comment (DzmitryShylovich) [inline modules/@angular/router/src/apply_redirects.ts]: `any` => `LoadedRouterConfig`

File: modules/@angular/router/src/apply_redirects.ts [hunk 5]
```diff
@@ -219,80 +226,80 @@ class ApplyRedirects {
...
+
+      if (childConfig.length === 0 && slicedSegments.length === 0) {
+        return of (new UrlSegmentGroup(consumedSegments, {}));
+      }
+
+      const expanded$ = this.expandSegment(
+          childInjector, segmentGroup, childConfig, slicedSegments, PRIMARY_OUTLET, true);
+      return map.call(
+          expanded$,
+          (cs: any) => new UrlSegmentGroup(consumedSegments.concat(cs.segments), cs.children));
```
Comment (DzmitryShylovich) [inline modules/@angular/router/src/apply_redirects.ts]: `any` => `UrlSegmentGroup`

File: modules/@angular/router/src/apply_redirects.ts [hunk 6]
```diff
@@ -219,80 +226,80 @@ class ApplyRedirects {
...
-            return map.call(this.configLoader.load(injector, route.loadChildren), (r: any) => {
-              (<any>route)._loadedConfig = r;
-              return r;
-            });
-          }
-        } else {
-          return canLoadFails(route);
+          return (<any>route)._loadedConfig ?
+              of ((<any>route)._loadedConfig) :
+              map.call(this.configLoader.load(injector, route), (r: any) => {
```
Comment (DzmitryShylovich) [inline modules/@angular/router/src/apply_redirects.ts]: `(r: any)` => `(c: LoadedRouterConfig)`

**Comment by vicb** [comment]
@DzmitryShylovich Thanks for the review !

File: modules/@angular/router/src/router_config_loader.ts [hunk 7]
```diff
@@ -29,14 +29,24 @@ export class LoadedRouterConfig {
...
 
   load(parentInjector: Injector, route: Route): Observable<LoadedRouterConfig> {
+    if (this.onLoadStartListener) {
+      this.onLoadStartListener(route);
+    }
+
     const moduleFactory$ = this.loadModuleFactory(route.loadChildren);
+
     return map.call(moduleFactory$, (factory: NgModuleFactory<any>) => {
+      if (this.onLoadEndListener) {
```
Comment (tbosch) [inline modules/@angular/router/src/router_config_loader.ts]: move after module creation.

**Review by tbosch — APPROVED** [review]

**Pull request closed without merging**